DynamicArray class

ryoung9550 opened the pull request:
This patch features an addition of a std::vector like class called DynamicArray.

ryoung9550 committed:
Added swap files to .gitignore

ryoung9550 committed:
Fixed include header to match the header name

ryoung9550 committed:
Added DynamicArray which is a basic implementation of std::vector

ryoung9550 committed:
Updated documentation

faheel commented on .gitignore:
```diff
@@ -4,3 +4,7 @@ bin
+
+
+## Swap Files
+*.swp
```
Are swap files related to Vim? If so then they get deleted automatically after you save and/or exit from the editor, and so they don't need to be added to gitignore.

ryoung9550 replied:
Swap files are temp files which vim makes when files are opened. I typically use multiple terminals with some files already open so these files show up in the git status. I'm not sure what the conventions are with .gitignore so if it doesn't fit the bill I'll remove it.

faheel replied:
Ah, thanks for the explanation. It's OK to add it to gitignore, but please change the comment to `Swap files for Vim`

faheel commented on C++/include/data_structures/dynamic_array.hpp:
```diff
@@ -0,0 +1,129 @@
+T& DynamicArray<T>::front()
+{
+	if (numOfItems > 0)
+		return itemList[0];
```
What about the `else` case? If `numOfItems` is 0, this will lead to an error. Similarly for `back()`.

faheel commented on C++/include/data_structures/dynamic_array.hpp:
```diff
@@ -0,0 +1,129 @@
+private:
+	unsigned arrSize;
+	unsigned numOfItems;
+	T* itemList;
```
- Use `size_t` instead of `unsigned`, as used in the STL.
- Please rename these data members so that they closely mimic those in the STL. Eg: `arrSize` can be renamed as `capacity` and `numOfItems` can be renamed as `size`. `itemList` can be renamed to just `items`.

faheel reviewed (changes requested):
This would be a good addition. Thanks!
However, there are a few changes that need to be made.
- Please use soft tabs that are 4 spaces long.
- For those function arguments whose value does not change, accept them as `const type& param` instead of `type param` to prevent multiple copies being made.
- For member functions that do not change any data members, mark them as `const`.
- Use `and` and `or` instead of `&&` and `||`.

ryoung9550 committed:
Fixed needed changes for pull request

Tab switched to 4 space soft tab.
Fixed member function which didn't edit member data so it is const safe.
Made arguments pass by const reference.
Fixed else case on DynamicArray<T>::front() and DynamicArray<T>::back()

faheel reviewed (commented):
Looks good to me, but its usage needs to be tested. It would be great if you could also add a unit test for DynamicArray. The instructions can be found in [UNIT_TESTS.md](https://github.com/faheel/Algos/blob/master/C%2B%2B/UNIT_TESTS.md).

ryoung9550 committed:
Create dynamic_array.test.cpp

faheel commented on C++/test/data_structures/dynamic_array.test.cpp:
```diff
@@ -0,0 +1,15 @@
+// Add your test cases here:
+#ifndef DYNAMIC_ARRAY
+
+
```
Looks like you forgot to add the tests :)

In case you aren't quite sure about what kind of tests to add or how to add them, you can take a look at the [unit test for DisjointSet](https://github.com/faheel/Algos/blob/master/C%2B%2B/test/data_structures/disjoint_set.test.cpp) for some inspiration.

faheel commented:
Quite a few changes have been made to the project, so please pull the changes from upstream to update your fork.

alxmjo commented:
Closing due to inactivity.
